### PR TITLE
switch internal/serialized representation from brackettree format to column-based export format, and dispense with token escaping

### LINKED ISSUE #76: Error when entering annotate mode when punctuation not separately tokenized from adjacent tokens
@mebwells was testing the new version and encountered the following error: 

`ValueError: Indices must occur at most once` 

This error is raised when punctuation isn't tokenized separately from adjacent tokens, as in:

`Not later than 270 days after the commencement of the study required by paragraph (1), the Secretary of Commerce shall, in consultation with the Secretary of Energy and the Director of the Office of Science and Technology Policy, submit to Congress a report on such study.`

It goes away with proper tokenization:

`Not later than 270 days after the commencement of the study required by paragraph ( 1 ) , the Secretary of Commerce shall , in consultation with the Secretary of Energy and the Director of the Office of Science and Technology Policy , submit to Congress a report on such study .`

### PR DESCRIPTION
**Pull request by bwaldon**
Handles escaping of sequences of form `(S)` where S is a sequence of characters that are not whitespace, `(`, or `)`. `(S)` can pose a problem for ptree conversion from brackettree strings (the result of calling `brackettree`), because `S` can be mis-analyzed as a terminal index. 

This solution escapes `(S)` with lookalike characters as `﹙S﹚` before ptree conversion. We also ensure that `﹙S﹚` sequences are unescaped in the tree tokenization on `__init__`, which is necessary in cases where a new `ActivedopTree` instance is created from string (using the `ActivedopTree.from_str()` classmethod).

### CODE REVIEW

**Comment by nschneid** [comment]
This is specific to the form `(S)`—what about other sequences containing `(` or `)`? Will they require escaping?

**Comment by bwaldon** [comment]
Not from what I can tell, based on my own testing -- things like `:)` or `:-(` aren't causing issues.

**Comment by nschneid** [comment]
Maybe it's just the open paren at the beginning that triggers the problem—what about `(=` or `((` or `()`?

**Comment by bwaldon** [comment]
Nice, in the most recent commit I have a new solution for this which handles those cases in addition to the case mentioned in https://github.com/nschneid/activedop/issues/76. An added advantage of this new solution is that it doesn't actually require us to introduce new escape tokens for parentheses. (Instead, we can escape with the sequence specified in PUNCT_ESCAPING).

**Comment by nschneid** [comment]
So how would these things be escaped in the new policy?

**Comment by bwaldon** [comment]
With our current way of defining `PUNCT_ESCAPING`, the new policy would be to escape any instance of `)` or `(` in a terminal with `-RRB-` or `-LRB-`, respectively. (E.g., `:)` -> `:-RRB-`). This escaping applies to ptrees only: `:t` and `:p` tags in CGELTree terminals will still feature the unescaped forms. The helper function `_apply_standard_labels` ensures that ptree labels associated with these terminals are well-formed. 

When it comes to a sequence like `:)`/`:-RRB-`, the annotator has the option to treat it as a `:p` tag of a CGELTree terminal or as a distinct CGELTree terminal. In the former case, `_apply_standard_labels` ensures that the label is `*-p`. These edits still need to be made 'manually' in the text box in edit mode: we don't (yet) support switching between punctuation/non-punct labels as a graphical edit.

**Comment by nschneid** [comment]
`:-RRB-` is pretty cryptic. But if the annotator doesn't see the ptree escaping then that's fine.

Another thought is that perhaps dopparser could alternately be called with a non-bracketed format that wouldn't require escaping?

**Comment by bwaldon** [comment]
It's very cryptic -- but yeah, the annotator will never actually see this. In the text window and the graphical tree, `-RRB-` and `-LRB-` sequences appear as `)` and `(`  respectively. 

We could also experiment with calling the dopparser w/ a non-bracketed format. We're currently dealing with the non-bracketed format because we translate CGELTrees to ParentedTree by first calling `.ptb()` method on a CGELTree (which returns a bracket-format tree string) and then converting the output of that method to a ParentedTree object. I think that in order to avoid dealing with the bracket format, we'd need a CGELTree method to convert to a non-bracket format.

**Comment by nschneid** [comment]
There's a command at the bottom of [this page](https://github.com/andreasvc/disco-dop/blob/master/docs/cli/parser.rst) suggesting that the discodop tool can convert between formats. I don't know how such conversions are supported at the API level. There are several [format options](https://github.com/andreasvc/disco-dop/blob/master/docs/fileformats.rst).

**Comment by bwaldon** [comment]
I see -- so maybe a CGELTree object method to convert to Negra export? I could get started on that! That way, we could convert to ParentedTree without a bracket-format intermediary. Or maybe what's really needed is a `toParentedTree` CGELTree object method (avoiding any intermediate format)

**Comment by nschneid** [comment]
It occurs to me that we have to think about two pipeline stages: parser training time and test time (annotation). Ideally the conversion from CGELBank format would be the same. Writing a CGELTree -> ParentedTree converter seems purest assuming discodop can convert ParentedTree <-> Negra export format.

**Comment by bwaldon** [comment]
Great, I'll investigate this!

**Comment by bwaldon** [comment]
> Writing a CGELTree -> ParentedTree converter seems purest assuming discodop can convert ParentedTree <-> Negra export format.

What if we sidestepped Negra export format conversion entirely? We currently rely on this format just once in `activedoptree.py` , as an intermediary in ParentedTree -> CGELTree conversion:

```
		tree_to_cgel = self._remove_punctuation_nodes(tree_copy) # tree_to_cgel: a ParentedTree object
		
		try:
			block = writetree(tree_to_cgel, non_punct_tokens, '1', 'export', comment='')
			block = io.StringIO(block)
			cgel_tree = next(load_as_cgel(block)) 
```

If we had both "load CGEL from ParentedTree" and "convert CGEL to ParentedTree" methods, we could avoid both Negra export-format *and* bracket-format intermediaries.

**Comment by nschneid** [comment]
Also for training? The training data needs to be in files right?

**Comment by bwaldon** [comment]
This would only apply to annotation (specifically, this would apply to how we convert to/from CGEL after user operations like graphical edits). We’d still need an output string format to write annotations to (and to train the parser). That format could be Negra export — it’s currently one of the formats written to the database (the other format is the native CGELBank .cgel format), and it’s supported by discodop for training. Right now, we’re training from bracket-format tree strings, which are converted from .cgel.

**Comment by nschneid** [comment]
If the discodop API already does ParentedTree <-> Negra, then it's just a question of which is simpler to implement: ParentedTree <-> CGELTree or Negra <-> CGELTree. I can't tell which is simpler without actually implementing it.

**Comment by bwaldon** [comment]
In https://github.com/nschneid/activedop/pull/82/commits/9a3f94d9da2c07740bb94e1767ee1b30df094791:

- CGELBank `.cgel` tree files are converted to Negra corpus 'export' format to train the dopparser -- `cgel2export` is a CLI utility in `app.py` which facilitates the conversion. 
- CGELBank `Tree` objects are converted to `ParentedTree` directly, using a new `cgel_to_ptree()`function defined in `activedoptree.py`. 
- `ParentedTree` objects are back-converted to CGELBank `Tree` objects as before (i.e., via conversion to export format). 
- `_apply_standard_labels` method of the `ActivedopTree` class is augmented with a few extra guardrails against producing illicit tree structures. 

Because we no longer use a bracket-format intermediary anywhere in this workflow, we need considerably less sequence escaping compared to before. (Notably, there is no longer any need to escape parentheses).

**Comment by nschneid** [comment]
So glad to say goodbye to all that escaping code!

File: activedoptree.py [hunk 1]
```diff
@@ -60,22 +100,22 @@ def __init__(self, ptree: ParentedTree, senttok: List[str], cgel_terminals = Non
...
+		self.ptree, self.senttok = cgel_to_ptree(self.cgel_tree)
 		# update canonicalized ptree with standardized labels
 		for i, subt in enumerate(self.ptree.subtrees(lambda t: t.height() == 2)):
 			subt.label = ptree_terminals_with_labels[i].label
 		# update cgel_tree with a set of terminals if provided
 		if cgel_terminals is not None:
 			self.cgel_tree.update_terminals(cgel_terminals, gaps=True, restore_old_cat=True, restore_old_func=True)
-	
+
 	def brackettreestr(self, pretty = False):
```
Comment (nschneid) [inline activedoptree.py:112]: Do we still need this?

Comment (nschneid) [reply activedoptree.py:112]: (I guess this question applies to all mentions of "bracket" in the code)

File: activedoptree.py [hunk 2]
```diff
@@ -158,8 +206,8 @@ def _ptree_to_cgel(self) -> ParentedTree:
 		Assumes that tree has been processed by apply_standard_labels().
 		Outputs both a CGELTree object and a ParentedTree object, the latter of which is the cleaned-up version of the input tree with a canonicalized position for punctuation preterminals/terminals.
```
Comment (nschneid) [inline activedoptree.py:207]: I see that this function returns a ParentedTree. Where does it output a CGELTree object? I'm confused about the name `_ptree_to_cgel` if the output is a ParentedTree.

Comment (bwaldon) [reply activedoptree.py:207]: Whoops, both the docstring and the type hint need to be updated here. The function outputs just a CGELTree object.

File: activedoptree.py (hunk 1 above)
Comment (bwaldon) [reply activedoptree.py:112]: We only need this for the legacy PTB edit mode, to write bracket tree strings to the text box. (We never use it to edit CGEL trees). I've been trying periodically to make sure that we still support editing in non-CGEL formats.

File: activedoptree.py (hunk 2 above)
Comment (nschneid) [reply activedoptree.py]: Looking at it some more, I am wondering if the return type annotation is wrong and should be `CGELTree`. I think it is stripping punctuation, converting to a ParentedTree, and then to a CGELTree. I wonder if there is some redundancy with the cgel2export endpoint in app.py.

**Comment by nschneid** [comment]
Remind me: Where exactly was the brackettree format used in annotation (not training) before this change? Was it the format produced by the parser, that had to be translated to the ParentedTree and/or gtree? Was it translated to/from CGEL format?

**Comment by bwaldon** [comment]
> Remind me: Where exactly was the brackettree format used in annotation (not training) before this change? Was it the format produced by the parser, that had to be translated to the ParentedTree and/or gtree? Was it translated to/from CGEL format?

It's how we translated from CGEL to ParentedTree. Before this change, we first called the CGELTree object's `.ptb()` method, then parsed the output of that method (a bracketed tree string) to get the ParentedTree object.

**Comment by nschneid** [comment]
OK, is this a valid summary?:

Adopts Negra export format to replace bracketed trees as a serialized representation. It is column-based, so tokens do not require escaping. (fixes #76 - escape sequences surrounded by parentheses)

Previously, in addition to being used for training data, the PTB-esque brackettree format was used as an intermediary between `CGELTree` and `ParentedTree` representations. (`ParentedTree` is the data structure used by dop-parser and the graphical tree renderer.)

Now, for CGEL annotation, translation is handled in `cgel2export` for training data, and `cgel_to_ptree()` (produces a `ParentedTree` object) for graphical rendering. The cgel library already implements activedopexport2cgel.py. (A legacy mode with bracketed text entry is still supported.)

**Comment by bwaldon** [comment]
Yes, looks good!

**Pull request merged**